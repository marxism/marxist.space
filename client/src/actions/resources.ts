import {
  fetchResources as _fetchResources,
  searchResources as _searchResources,
  filterResources as _filterResources,
} from '../util/api'

export const FETCH_RESOURCES_REQUEST = 'FETCH_RESOURCES_REQUEST'
export const FETCH_RESOURCES_SUCCESS = 'FETCH_RESOURCES_SUCCESS'
export const FETCH_RESOURCES_ERROR = 'FETCH_RESOURCES_ERROR'

const fetchResourcesRequest = { type: FETCH_RESOURCES_REQUEST }
const fetchResourcesSuccess = (resources) => ({
  type: FETCH_RESOURCES_SUCCESS,
  resources,
})
const fetchResourcesError = (error) => ({ type: FETCH_RESOURCES_ERROR, error })

export const fetchResources = () => async (dispatch) => {
  dispatch(fetchResourcesRequest)
  try {
    const resources = await _fetchResources()
    dispatch(fetchResourcesSuccess(resources))
  } catch (error) {
    dispatch(fetchResourcesError(error))
  }
}

export const FILTER_RESOURCES_REQUEST = 'FILTER_RESOURCES_REQUEST'
export const FILTER_RESOURCES_SUCCESS = 'FILTER_RESOURCES_SUCCESS'
export const FILTER_RESOURCES_ERROR = 'FILTER_RESOURCES_ERROR'

const filterResourcesRequest = (tags: string[]) => ({
  type: FILTER_RESOURCES_REQUEST,
  payload: tags,
})
const filterResourcesSuccess = (resources) => ({
  type: FILTER_RESOURCES_SUCCESS,
  resources,
})
const filterResourcesError = (error) => ({
  type: FILTER_RESOURCES_ERROR,
  error,
})

export const filterResources = (tags: string[]) => async (dispatch) => {
  dispatch(filterResourcesRequest(tags))
  try {
    const resources = await _filterResources(tags)
    dispatch(filterResourcesSuccess(resources))
  } catch (error) {
    dispatch(filterResourcesError(error))
  }
}

export const SEARCH_RESOURCES_REQUEST = 'SEARCH_RESOURCES_REQUEST'
export const SEARCH_RESOURCES_SUCCESS = 'SEARCH_RESOURCES_SUCCESS'
export const SEARCH_RESOURCES_ERROR = 'SEARCH_RESOURCES_ERROR'

const searchResourcesRequest = (text: string) => ({
  type: SEARCH_RESOURCES_REQUEST,
  payload: text,
})
const searchResourcesSuccess = (resources) => ({
  type: SEARCH_RESOURCES_SUCCESS,
  resources,
})
const searchResourcesError = (error) => ({
  type: SEARCH_RESOURCES_ERROR,
  error,
})

export const searchResources = (text: string) => async (dispatch) => {
  dispatch(searchResourcesRequest(text))
  try {
    const resources = await _searchResources(text)
    dispatch(searchResourcesSuccess(resources))
  } catch (error) {
    dispatch(searchResourcesError(error))
  }
}
