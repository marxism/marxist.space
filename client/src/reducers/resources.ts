import {
  FETCH_RESOURCES_REQUEST,
  FETCH_RESOURCES_SUCCESS,
  FETCH_RESOURCES_ERROR,
  FILTER_RESOURCES_REQUEST,
  FILTER_RESOURCES_SUCCESS,
  FILTER_RESOURCES_ERROR,
  SEARCH_RESOURCES_REQUEST,
  SEARCH_RESOURCES_SUCCESS,
  SEARCH_RESOURCES_ERROR,
} from '../actions/resources'

const initialState = { isFetching: false, items: [] }

export default (state = initialState, action) => {
  switch (action.type) {
    case FETCH_RESOURCES_REQUEST:
    case FILTER_RESOURCES_REQUEST:
    case SEARCH_RESOURCES_REQUEST:
      return { ...state, isFetching: true, resource: null, newResource: null }
    case FETCH_RESOURCES_SUCCESS:
    case FILTER_RESOURCES_SUCCESS:
    case SEARCH_RESOURCES_SUCCESS:
      return { ...state, isFetching: false, items: action.resources }
    case FETCH_RESOURCES_ERROR:
    case FILTER_RESOURCES_ERROR:
    case SEARCH_RESOURCES_ERROR:
      return { ...state, isFetching: false }

    default:
      return state
  }
}
