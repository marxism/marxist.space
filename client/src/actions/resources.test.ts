/* eslint-env jest */

import * as resourcesActions from './resources'

describe('resources actions', () => {
  test('fetchResourcesRequest', () => {
    expect(resourcesActions.fetchResourcesRequest).toMatchSnapshot()
  })

  test('fetchResourcesSuccess', () => {
    expect(
      resourcesActions.fetchResourcesSuccess(['one', 'two'])
    ).toMatchSnapshot()
  })

  test('fetchResourcesError', () => {
    expect(resourcesActions.fetchResourcesError('oh no')).toMatchSnapshot()
  })

  test('filterResourcesRequest', () => {
    expect(resourcesActions.filterResourcesRequest(['one'])).toMatchSnapshot()
  })

  test('filterResourcesSuccess', () => {
    expect(
      resourcesActions.filterResourcesSuccess(['one', 'two'])
    ).toMatchSnapshot()
  })

  test('filterResourcesError', () => {
    expect(resourcesActions.filterResourcesError('oh no')).toMatchSnapshot()
  })

  test('searchResourcesRequest', () => {
    expect(resourcesActions.searchResourcesRequest('one')).toMatchSnapshot()
  })

  test('searchResourcesSuccess', () => {
    expect(
      resourcesActions.searchResourcesSuccess(['one', 'two'])
    ).toMatchSnapshot()
  })

  test('searchResourcesError', () => {
    expect(resourcesActions.searchResourcesError('oh no')).toMatchSnapshot()
  })
})
