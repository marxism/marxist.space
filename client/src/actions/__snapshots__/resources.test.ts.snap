// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resources actions fetchResourcesError 1`] = `
Object {
  "error": "oh no",
  "type": "FETCH_RESOURCES_ERROR",
}
`;

exports[`resources actions fetchResourcesRequest 1`] = `
Object {
  "type": "FETCH_RESOURCES_REQUEST",
}
`;

exports[`resources actions fetchResourcesSuccess 1`] = `
Object {
  "resources": Array [
    "one",
    "two",
  ],
  "type": "FETCH_RESOURCES_SUCCESS",
}
`;

exports[`resources actions filterResourcesError 1`] = `
Object {
  "error": "oh no",
  "type": "FILTER_RESOURCES_ERROR",
}
`;

exports[`resources actions filterResourcesRequest 1`] = `
Object {
  "payload": Array [
    "one",
  ],
  "type": "FILTER_RESOURCES_REQUEST",
}
`;

exports[`resources actions filterResourcesSuccess 1`] = `
Object {
  "resources": Array [
    "one",
    "two",
  ],
  "type": "FILTER_RESOURCES_SUCCESS",
}
`;

exports[`resources actions searchResourcesError 1`] = `
Object {
  "error": "oh no",
  "type": "SEARCH_RESOURCES_ERROR",
}
`;

exports[`resources actions searchResourcesRequest 1`] = `
Object {
  "payload": "one",
  "type": "SEARCH_RESOURCES_REQUEST",
}
`;

exports[`resources actions searchResourcesSuccess 1`] = `
Object {
  "resources": Array [
    "one",
    "two",
  ],
  "type": "SEARCH_RESOURCES_SUCCESS",
}
`;
